README
WWW-Scripter-Plugin-JavaScript-0.008a (alpha)

A plugin for  WWW::Scripter  that provides JavaScript support

This distribution contains:
WWW::Scripter::Plugin::JavaScript     0.008 — The JavaScript plugin
WWW::Scripter::Plugin::JavaScript::JE 0.009 — A back end for JE


RECENT CHANGES

(See the Changes file for the long version.)

0.008a
-----
• LWP 6 supports perl 5.8.7 once  more,  so  we  no  longer  require
  Bundle::LWP5_837.

• The JE back end now binds window methods properly, eliminating err-
  ors with JE::Object::Proxy objects  (RT #79855).  Thanks to Nathan
  Chaney for the report.

0.008
-----
• The JavaScript plugin now tells back ends when it has finished with
  them, so they have a chance to break circular references.
• The  JE  back end takes advantage of that,  to avoid  memory  leaks.

0.007
-----
Makefile.PL  now requires  Bundle::LWP5_837  instead of  LWP  on perl
5.8.7 and lower, since LWP 6 does not support 5.8.7.

0.006
-----
• The JE back end’s event2sub method  (that parses HTML  event  attri-
  butes)  now sets  $@  instead of dying if there is a  syntax  error.


INSTALLATION

The easiest way to install this module is to use the CPAN module or
the cpan script:

    [sudo] perl -MCPAN -e "install WWW::Scripter::Plugin::JavaScript"

    [sudo] cpan WWW::Scripter::Plugin::JavaScript

Or you can use the following:

    perl Makefile.PL
    make
    make test
    [sudo] make install


DEPENDENCIES

This module requires:

- perl 5.8.4 or higher

- JE 0.056 or higher

- HTML::DOM 0.032 or higher

- CSS::DOM

- WWW::Scripter 0.022 or higher

- Hash::Util::FieldHash::Compat

- URI

- LWP 5.815 or higher


DOCUMENTATION

After installing, you can find documentation for these modules with
the perldoc command.

    perldoc WWW::Scripter::Plugin::JavaScript
    perldoc WWW::Scripter::Plugin::JavaScript::JE

Or try using man (it's faster, in my experience):

    man WWW::Scripter::Plugin::JavaScript
    man WWW::Scripter::Plugin::JavaScript::JE


COPYRIGHT AND LICENCE

Copyright © 2009-11 Father Chrysostomos

This program is free software; you may redistribute it and/or modify
it under the same terms as perl.


Changes
Revision history for WWW::Scripter::Plugin::JavaScript

0.008a  21 October, 2012
        LWP 6 supports perl 5.8.7 once more,  so we no longer require
        Bundle::LWP5_837.
        JE back end (0.009):
        • Window methods are now bound properly,  eliminating  errors
          with  JE::Object::Proxy  objects  (RT  #79855).  Thanks  to
          Nathan Chaney for the report.

0.008   4 December, 2011
        The JavaScript plugin:
        • The JavaScript plugin now calls  ->destroy  on the back end
          when it has finished with it.  This gives the  back  end  a
          chance to destroy any circular  references  it  might  have.
        JE back end:
        • It now takes advantage of JE’s new JE::Destroyer module, to
          eliminate circular references.

0.007   26 March, 2011
        Makefile.PL now requires Bundle::LWP5_837 instead of LWP on
        perl 5.8.7 and lower,  since  LWP  6 does not support 5.8.7.
        The JavaScript plugin:
        • An internal change that should not affect the  interface:
          The  JavaScript  plugin no longer implements  the  screen
          object. WWW::Scripter now implements it, as should always
          have been the case.

0.006   18 January, 2010
        The JavaScript plugin:
        • The docs have  been  updated  to  clarify  that  back  ends’
          event2sub  and  eval  methods must  set  $@  on  error  and
          return normally.
        • Some more doc changes  to  reflect  the  current  state  of
          SpiderMonkey support.
        JE back end (0.007):
        • event2sub  now sets  $@  instead of dying if there is a syn-
          tax error.

0.005   14 February, 2010
        The JavaScript plugin:
        • Incompatible changes:
          • set  now expects a  WWW::Scripter  object  as  the  first
            argument.
          • new_function,  which used to offer no more  functionality
            than  set,  now adds the function to all JS  environments,
            including future ones (a bit like bind_classes).
        • New back_end method
        • The eval method is now public (it was there all along,  but
          the docs were removed  in  the  conversion  from  the  Mech
          plugin  to  the  Scripter  plugin;  the  interface  changed
          slightly in the process).
        • JE is now the default back end. SpiderMonkey used to be the
          back end,  but there was never one available,  so that made
          no difference.  Now there is a SpiderMonkey back  end  (not
          included),  but it is so full of bugs we need to wait a bit
          before flipping the switch back.
        JE back end (0.006):
        • A piece of JavaScript code  calling  a  method  on  another
          global object no longer dies.  This was broken  by  0.004’s
          multiple-window clean-up.
        • Modified to make use of the plugin’s  new  back_end  method,
          to reduce memory usage.

0.004   25 October, 2009
        The JavaScript Plugin:
        • Scripts in frames now run in the context of the frame,  not
          the top-level window.
        JE back end (0.005):
        • We now explicitly require version 0.032 of HTML::DOM::Inter-
          face.  (This should have been added  in  the  previous  ver-
          sion.)

0.003   9 October, 2009
        The JavaScript plugin:
        • Updated to work with  HTML::DOM  0.032’s new event  handler
          model. This fixes various bugs involving event handler prop-
          erties (‘theForm.onsubmit’).
        • Calling  WWW::Scripter’s  eval  method would run  the  Java-
          Script code always in the main window,  if when called from
          the WWW::Scripter object corresponding to a frame. This has
          been fixed. This is part of the solution to the frames prob-
          lem mentioned below.
        •‘require WWW::Scripter::Plugin::JavaScript’  no longer  dies
          if WWW::Scripter is not already loaded.
        JE back end (0.004):
        • We now support those utterly useless  HTML-generating  meth-
          ods of string objects that have  been  part  of  JavaScript
          since the very beginning,  and which have been added solely
          for completeness’ sake.
        • We now override JE’s exists method, so that collection prop-
          erties of the window object appear in scope and can be seen
          by hasOwnProperty.
        • Likewise updated to work with the new event  handler  model.
        • The scope  of  an  event  handler  now  includes  the  docu-
          ment object.
        • For those cases where HTML::DOM  provides UTF-16 methods in
          addition to the usual ones,  JavaScript now uses the UTF-16
          versions.  (This makes  textNode.data.length  equivalent to
          textNode.length.)
        • It used to be that frames with no scripts in them that were
          accessed by JavaScript code in the main window would  share
          the global  object  with  the  main  window.  Also,  window
          objects,  if referenced by JavaScript variables in  another
          frame,  would continue to refer to the same  global  object
          when the frame was navigated.  This mess  has  been  mostly
          cleaned up. There are a few rough edges that will be ironed
          out in a future release.

0.002c  27 August, 2009
        JE back end (0.003):
        • This release simply fixes a test  to  check  for  JE  0.035
          instead of  JE  0.034  before deciding not to skip  a  test.
          (The version has been changed as  the  module’s  docs  have
          changed.)

0.002b  25 August, 2009
        JE back end:
        • It now turns on  JE’s HTML mode  (to support HTML  comments
          embedded in JavaScript).  Version 0.034 of  JE  is required
          for this feature to work.

0.002a  16 August, 2009
        HTML::DOM 0.027 and WWW::Scripter 0.004  are now required.  A

Makefile.PL
#!/usr/bin/perl

BEGIN{ require 5.006; }
use ExtUtils::MakeMaker;

# I’m keeping these separate, in case these modules should ever be dis-
# tributed individually:
%js_prq = (
	CSS::DOM::Interface           => 0,
	Encode                        => 0,
	Hash::Util::FieldHash::Compat => 0,
	HTML'DOM'EventTarget          => 0.032, # new event handler model
	HTML::DOM::Interface          => 0,
	Scalar::Util                  => 0,
	strict                        => 0,
	URI::Escape                   => 0,
	warnings                      => 0,
	WWW'Scripter                  => 0.022, # screen
# for testing:
	lib          => 0,
	Test::More   => 0,
	URI::file    => 0,
);
#if($] < 5.0080071 && !eval 'use LWP::UserAgent 5.815; 1') {
# $js_prq{'Bundle::LWP5_837'} = 0;
#}
#else {
 $js_prq{'LWP::UserAgent'} = 5.815; # handlers
#}

%je_prq = (
	Carp                          => 0,
	Hash::Util::FieldHash::Compat => 0,
	HTML::DOM::Interface          => 0.032, # UTF16 export
	JE                            =>  .038, # call_with
	JE::Destroyer                 => 0,
	Scalar::Util                  => 1.09, # refaddr
	strict                        => 0,
	warnings                      => 0,
# for testing:
	HTML'DOM     => 0.027, # document.write and node lists
	lib          => 0,
	Test::More   => 0,
	URI::file    => 0,
	WWW'Scripter =>  .016, # event2sub and $@
#	WWW'Scripter'Plugin'JavaScript => 0,
);
eval{require HTML::DOM and VERSION HTML::DOM == 0.034}
 and $je_prq{HTML::DOM} = 0.035; # getElementById bug

for(\(%js_prq, %je_prq)) {
	for $k(keys %$_) {
		$prq{$k} = $$_{$k}
			unless exists $prq{$k} and $$_{$k} < $prq{$k};
	}
}

WriteMakefile(
    NAME                => 'WWW::Scripter::Plugin::JavaScript',
    AUTHOR              => 'Father Chrysostomos <sprout & cpan . org>',
#    VERSION_FROM        => 'lib/WWW/Scripter/Plugin/JavaScript.pm',
    VERSION             => "0.008a",
    ABSTRACT_FROM       => 'lib/WWW/Scripter/Plugin/JavaScript.pm',
    PL_FILES            => {},
    PREREQ_PM           => \%prq,
    dist               => { COMPRESS => 'gzip -9f', SUFFIX => 'gz',
                            TAR => 'COPYFILE_DISABLE=1 tar' },
    clean              => { FILES => 'WWW-Scripter-Plugin-JavaScript-*' },
);


lib/WWW/Scripter/Plugin/JavaScript.pm
package WWW::Scripter::Plugin::JavaScript;

use strict;   # :-(
use warnings; # :-(

use Encode 'decode_utf8';
use LWP'UserAgent 5.815;
use Scalar::Util qw'weaken';
use URI::Escape 'uri_unescape';
use Hash::Util::FieldHash::Compat 'fieldhash';
use WWW::Scripter 0.022; # screen

our $VERSION = '0.008';

# Attribute constants (array indices)
sub mech() { 0 }
sub jsbe() { 1 } # JavaScript back-end (field hash of objects, keyed
sub benm() { 2 } # Back-end name  # by document)
sub init_cb() { 3 } # callback routine that's called whenever a new js
                    # environment is created
sub alert()   { 4 }
sub confirm() { 5 }
sub prompt()  { 6 }
sub cb() { 7 } # class bindings
sub tmout() { 8 } # timeouts
sub f()       { 9 } # functions
sub g()        { 10 } # guard objects for back ends, to destroy
                      # them forcibly

{no warnings; no strict;
undef *$_ for qw/mech jsbe benm init_cb g cb
              f alert confirm prompt tmout/} # These are PRIVATE constants!

sub init {

	my ($package, $mech) = @_;

	my $self = bless [$mech], $package;
	weaken $self->[mech];

	$mech->script_handler( default => $self );
	$mech->script_handler(
	 qr/(?:^|\/)(?:x-)?(?:ecma|j(?:ava)?)script[\d.]*\z/i => $self
	);

	$mech->set_my_handler(request_preprepare => sub {
		my($request,$mech) = @_;
		$self->eval(
		 $mech, decode_utf8 uri_unescape opaque {uri $request}
		);
		$@ and $mech->warn($@);
		WWW'Scripter'abort;
	}, m_scheme => 'javascript');

	# stop closures from preventing destruction
	weaken $mech;
	my $life_raft = $self;
	weaken $self;

	$self;
}

sub options {
	my $self = shift;
	my %opts = @_;

	my $w;
	for(keys %opts) {
		if($_ eq 'engine') {
			if($self->[jsbe] &&
			   $self->[benm] ne $opts{$_}
			) {
			    $self->[mech]->die(
			        "Can't set JavaScript engine to " .
			        "'$opts{$_}' since $self->[benm] is " .
			        "already loaded.");;
			}
			$self->[benm] = $opts{$_};;
		}
		elsif($_ eq 'init') {
			$self->[init_cb] = $opts{$_};
		}
		else {
			$self->[mech]->die(
			    "JavaScript plugin: Unrecognized option '$_'"
			);
		}
	}
}

sub eval {
	my($plugin,$mech,$code,$url,$line,$inline) = @_;

	if(
	 $code =~ s/^(\s*)<!--[^\cm\cj\x{2028}\x{2029}]*(?x:
	         )(?:\cm\cj?|[\cj\x{2028}\x{2029}])//
	) {
		$line += 1 + (()= $1 =~ /(\cm\cj?|[\cj\x{2028}\x{2029}])/g)
	}
	$code =~ s/-->\s*\z//;
		
	my $be = $plugin->back_end($mech);

	$be->eval($code, $url, $line);
}

sub event2sub {
		my($self,$mech,$elem,undef,$code,$url,$line) = @_;

		$self->
		  back_end($mech)->event2sub($code,$elem,$url,$line);
}

# We have to associate each JS environment with a response object. While
# writing this logic, I initially tried to use the document, but not all
# URLs have documents (e.g., plain text files).
sub back_end {
	my $self = shift;
ref $_[0] or require Carp, Carp'cluck();
	my $res = (my $w = shift)->res;
	return $self->[jsbe]{$res}
	 if ($self->[jsbe] ||= &fieldhash({}))->{$res};
	
	if(!$self->[benm]) {
# When wspjssm is stable enough, these lines can be uncommented:
#	    # try this one first, since it's faster:
#	    eval{require WWW::Scripter::Plugin::JavaScript::SpiderMonkey};
#	    if($@) {
	        require 
	            WWW::Scripter::Plugin::JavaScript::JE;
                $self->[benm] = 'JE'
#            }
#	    else { $self->[benm] = 'SpiderMonkey' };
	}
	else {
		require "WWW/Scripter/Plugin/JavaScript/" .
			"$$self[benm].pm";
	}

	($self->[g] ||= &fieldhash({}))->{$res}
	 = new WWW'Scripter'Plugin'JavaScript'Guard
	my $back_end = $self->[jsbe]{$res}
	 = "WWW::Scripter::Plugin::JavaScript::$$self[benm]" -> new( $w );
	require HTML::DOM::Interface;
	require CSS::DOM::Interface;
	for ($back_end) {
		for my $class_info( $self->[mech]->class_info ) {
		 $_->bind_classes($class_info) ;
		}
		for my $__(@{$self->[cb]||[]}){
			$_->bind_classes($__)
		}
		for my $__(@{$self->[f]||[]}){
			$_->new_function(@$__)
		}
	} # for $back_end;
	{ ($self->[init_cb]||next)->($w); }
	weaken $self; # closures
	return $back_end;
}

sub bind_classes {
	my $plugin = shift;
	push @{$plugin->[cb]}, $_[0];
	if($plugin->[jsbe]) {
		$_ && $_->bind_classes($_[0])
		 for values %{ $plugin->[jsbe] };
	}
}

sub set { shift->back_end( shift )->set(@_) }

sub new_function {
	my $plugin = shift;
	push @{$plugin->[f]}, \@_;
	if($plugin->[jsbe]) {
		$_ && $_->new_function(@_)
		 for values %{ $plugin->[jsbe] };
	}
}


# ~~~ This is experimental. The purposed for this is that code that relies
#     on a particular version of a JS back end can check to see which back
#     end is being used before doing Foo->VERSION($bar). The problem with
#     it is that it returns nothing unless the JS environment has already
#     been loaded. If we have it start the JS engine, we may load it and
#     then not use it.
sub engine { shift->[benm] }


package WWW::Scripter::Plugin::JavaScript::Guard;

sub new { bless \(my $object = pop) }
DESTROY { eval { ${$_[0]}->destroy } }


# ------------------ DOCS --------------------#

1;


=head1 NAME

WWW::Scripter::Plugin::JavaScript - JavaScript plugin for WWW::Scripter

=head1 VERSION

Version 0.008 (alpha)

=head1 SYNOPSIS

  use WWW::Scripter;
  $w = new WWW::Scripter;
  
  $w->use_plugin('JavaScript');
  $w->get('http://www.cpan.org/');
  $w->get('javascript:alert("Hello!")'); # prints Hello!
  
  $w->use_plugin(JavaScript =>
          engine  => 'SpiderMonkey',
          init    => \&init, # initialisation function
  );                         # for the JS environment
  
=head1 DESCRIPTION

This module is a plugin for L<WWW::Scripter> that provides JavaScript
capabilities (who would have guessed?).

To load the plugin, just use L<WWW::Scripter>'s C<use_plugin> method:

  $w = new WWW::Scripter;
  $w->use_plugin('JavaScript');

You can pass options to the plugin via the C<use_plugin> method. It takes
hash-style arguments and they are as follows:

=over 4

=item engine

Which JavaScript back end to use. Currently, the only two back ends
available are L<JE>, a pure-Perl JavaScript interpreter, and
L<WWW::Scripter::Plugin::SpiderMonkey> (that back end is bundled
separately). The SpiderMonkey back end is just a proof-of-concept as of
July, 2010, but may become the default in a future version. JE is now the
default.

If this option is
not specified, either JE or SpiderMonkey will be used, whichever is
available. It is possible to
write one's own bindings for a particular JavaScript engine. See below,
under L</BACK ENDS>. 

=item init

Pass to this option a reference to a subroutine and it will be run every
time a new JavaScript environment is initialised. This happens after the
functions above have been created. The first argument will
be the WWW::Scripter object. You can use this, for instance, 
to make your
own functions available to JavaScript.

=back

=head1 METHODS

L<WWW::Scripter>'s C<use_plugin> method will return a plugin object. The
same object can be retrieved via C<< $w->plugin('JavaScript') >> after the
plugin is loaded. The same plugin object is used for every page and frame,
and for every new window derived from the WWW::Scripter object. The
following methods can be called on that object:

=over 4

=item eval

This evaluates the JavaScript code passed to it. The WWW::Scripter object
is the first argument; the string of code the second. You can optionally
pass
two more arguments: the file name or URL, and the first line number.

This method sets C<$@> and returns C<undef> if there is an error.

=item set

Sets the named variable to the value given. The first argument is the
WWW::Scripter object. The last argument is the value. The intervening
arguments are the names of properties, so if you want to assign to a
property of a property ... of a global property, you can pass each property
name separately like this:

  $w->plugin('JavaScript')->set(
      $w, 'document', 'location', 'href' => 'http://www.perl.org/'
  );

=item new_function

This creates a new global JavaScript function out of a coderef. This
function is added to every JavaScript environment the plugin has access to. Pass the WWW::Scripter object as the first argument, the 
name as
the second and the code ref as the third.

=item bind_classes

Instead of using this method, you might consider L<WWW::Scripter>'s
C<class_info> method, which is more general-purpose (it applies also to
whatever other scripting languages might be available).

With this you can bind Perl classes to JavaScript, so that JavaScript can
handle objects of those classes. These class bindings will persist from one
page to the next.

You should pass a hash ref that has the
structure described in L<HTML::DOM::Interface>, except that this method
also accepts a C<< _constructor >> hash element, which should be set to the
name of the method to be called when the constructor function is called
within JavaScript; e.g., C<< _constructor => 'new' >>.

=item back_end

This returns the back end corresponding to the WWW::Scripter object passed
to it, creating it if necessary. This is intended mostly for back ends
themselves to use, for accessing frames, etc.

=back

=head1 FEATURES AVAILABLE TO JAVASCRIPT

The members of the HTML DOM that are available depend on the versions of
L<HTML::DOM> and L<CSS::DOM> installed. See L<HTML::DOM::Interface> and
L<CSS::DOM::Interface>.

For a list of the properties of the window object, see 
L<WWW::Scripter>.

=head1 BACK ENDS

A back end has to be in the WWW::Scripter::Plugin::JavaScript:: name
space. It will be C<require>d by this plugin implicitly when its name is
passed to the C<engine> option.

The following methods must be implemented:

=head2 Class methods

=over 4

=item new

This method is passed a window (L<WWW::Scripter>)
object.

It has to create a JavaScript environment, in which the global object
delegates to the window object for the members listed in 
L<C<%WWW::Scripter::WindowInterface>| WWW::Scripter::WindowInterface/THE C<%WindowInterface> HASH>.

When the window object or its frames collection (WWW::Scripter::Frames
object) is passed to the JavaScript 
environment, the global
object must be returned instead.

This method can optionally create C<window>, C<self> and C<frames>
properties
that refer to the global object, but this is not necessary. It might make
things a little more efficient.

Finally, it has to return an object that implements the interface below.

The back end has to do some magic to make sure that, when the global object
is passed to another JS environment, references to it automatically point
to a new global object when the user (or calling code) browses to another
page.

For instance, it could wrap up the global object in a proxy object
that delegates to whichever global object corresponds to the document.

=back

=head2 Object Methods

=over 4

=item eval

This should accept up to three arguments: a string of code, the file name
or URL, and the first line number.

It must set C<$@> and return C<undef> if there is an error.

=item new_function

=item set

=item bind_classes

These correspond to those 
listed above for
the plugin object. Unlike the above, though, this C<set> is not passed a
window as its first argument. Also, C<bind_classes> and C<new_function> are
only expected to act on a single JavaScript environment. The plugin's own
methods of the same names make sure every JavaScript environment's methods
are called.

C<new_function> must also accept a third argument, indicating the return
type. This (when specified) will be the name of a JavaScript function that
does the type conversion. Only 'Number' is used right now.
This requirement may be removed before version 1.

=item event2sub ($code, $elem, $url, $first_line)

This method needs to turn the
event handler code in C<$code> into an object with a C<call_with> method
and then return it. That object's C<call_with>
method will be
called with the event target and the event
object as its two arguments. Its return 
value, if
defined, will be used to determine whether the event's C<preventDefault>
method is called.

The function's scope must contain the following objects: the global object,
the document, the element's form (if there is one) and the element itself.

If the C<$code> could not be compiled, this method must set C<$@> and
return C<undef>, just like C<eval>.

=item define_setter

This will be called
with a list of property names representing the 'path' to the property. The
last argument will be a coderef that must be called with the value assigned
to the property.

B<Note:> This is actually not used right now. The requirement for this may
be removed some time before version 1.

=head1 PREREQUISITES

perl 5.8.4 or higher

HTML::DOM 0.032 or higher

JE 0.056 or later (when the SpiderMonkey binding is stable enough it will 
become optional)

CSS::DOM

WWW::Scripter 0.022 or higher

URI

Hash::Util::FieldHash::Compat

LWP 5.815 or higher

=head1 BUGS

=for comment
(See also L<WWW::Scripter::Plugin::JavaScript::JE/Bugs>.)

There is currently no system in place for preventing pages from different
sites from communicating with each other.

To report bugs, please e-mail the author.

=head1 AUTHOR & COPYRIGHT

Copyright (C) 2009-11 Father Chrysostomos
<C<< join '@', sprout => join '.', reverse org => 'cpan' >>E<gt>

This program is free software; you may redistribute it and/or modify
it under the same terms as perl.

=head1 SEE ALSO

=over 4

=item -

L<WWW::Scripter>

=item -

L<HTML::DOM>

=item -

L<JE>

=item -

L<WWW::Scripter::Plugin::JavaScript::SpiderMonkey>

=item -

L<JavaScript.pm|JavaScript>

=item -

L<JavaScript::SpiderMonkey>

=item -

L<WWW::Mechanize::Plugin::JavaScript> (the original version of this module)

=back


lib/WWW/Scripter/Plugin/JavaScript/JE.pm
package WWW::Scripter::Plugin::JavaScript::JE;

use strict;   # :-(
use warnings; # :-(

use Carp 'croak';
use Hash::Util::FieldHash::Compat 'fieldhash';
use HTML::DOM::Interface 0.032 ':all'; # for the constants (0.032
use JE'Destroyer;                      # for UTF16)
use JE 0.038; # call_with
use Scalar::Util 1.09 qw'weaken refaddr';

our $VERSION = '0.008';
our @ISA = 'JE';

fieldhash my %parathia;

# No need to implement eval since JE's method
# is sufficient

my @types;
$types[BOOL] = Boolean =>;
$types[STR ] = DOMString  =>;
$types[NUM ] = Number  =>;
$types[OBJ ] = null    =>;

sub new {
	my $self = SUPER::new{shift} html_mode => 1;
	weaken( $parathia{$self} = my $parathi = shift );
	
	my $i = \%WWW'Scripter'WindowInterface;
	for(grep !/^_/ && $$i{$_} & METHOD, =>=> keys %$i) {
		my $method = $_;
		my $type = $$i{$_};
		$self->new_method($_ => sub {
			my $parathi = $parathia{my $self = shift};
			# undocumented JE method:
			$self->_cast(
				scalar $parathi->$method(@_),
				$types[$type&TYPE]
			);
		});
	}
	for(grep !/^_/ && !($$i{$_}&METHOD) =>=> keys %$i) {
		my $name = $_;
		next if $name =~ /^(?:window|self)\z/; # for efficiency
		my $type = $$i{$_};
		if ($type & READONLY) {
			$self->prop({
				name => $_,
				readonly => 1,
				fetch => sub {
					my $self = shift;
					$parathia{$self}->$name;
					$self->_cast(
						scalar
						  $parathia{$self}->$name,
						$types[$type&TYPE]
					);
				},
			});
		}
		else {
			$self->prop({
				name => $_,
				fetch => sub {
					my $self = shift;
					$self->_cast(
						scalar
						  $parathia{$self}->$name,
						$types[$type&TYPE]
					);
				},
				store => sub {
					my $self = shift;
					$self->_cast(
						scalar
						  $parathia{$self}
						    ->$name(shift),
						$types[$type&TYPE]
					);
				},
			});
		}
	}

	# ~~~ This is *such* a hack! If anyone wants to help me fix JE’s
	#     screwed up type-conversion system (which makes my head hurt),
	#     please let me know.
	$self->new_function("DOMString", sub {
		if(ref($_[0]) =~ /^JE::(?:Null|Undefined)\z/) {
			return $_[0]->global->null;
		}
		return $_[0]->to_string;
	});

	$self->bind_class(
		package => 'WWW::Scripter',
		wrapper => sub {
			my ($self, $window) = @_;

			# ~~~ This needs to be modified to create a special
			#     restrictive wrapper if the $window has a
			#     different origin.
			# Fetch the cached JavaScript  environment  corres-
			# ponding to the WWW::Scripter object.
#warn $window, " ", $window->response, " ", $window->uri;
			refaddr
			 $window->plugin("JavaScript")->back_end($window)
			 == refaddr $self
			  and return $self;
			(__PACKAGE__."::Proxy")->new($window)
		},
	);
	$self->bind_class(
		package => 'WWW::Scripter::Frames',
		wrapper => sub {
			my($self,$frames) = @_;
			return $self->upgrade($frames->window);
		},
	);
	# ~~~ We also need a 'JE' wrapper, that will create a special
	#     objcet that delegates to the JS environment currently
	#     belonging to the window.

	for my $p($self->{String}{prototype}) {
		for my $v(
		 ['anchor','a','name'],
		 ['fontcolor','font','color'],
		 ['fontsize','font','size'],
		 ['link','a','href'],
		) {
			$p->new_method(
			 $$v[0],
			 sub{
			  qq'<$$v[1] $$v[2]="'
			  . (defined $_[1] ? $_[1] : 'undefined')
			  . qq'">$_[0]</$$v[1]>'
			 }
			)
		}
		for my $v('big', 'blink', 'small','strike','sub','sup') {
			$p->new_method( $v, sub{"<$v>$_[0]</$v>"} )
		}
		for my $v(['bold','b'],['fixed','tt'],['italics','i'],) {
			$p->new_method(
			 $$v[0],
			 sub{"<$$v[1]>$_[0]</$$v[1]>"}
			);
		}
	}

	# for speed:
	$self->prop('frames' => $self);
	$self->prop('window' => $self);
	$self->prop('self' => $self);
}

sub prop {
	my $self = shift;
	return $self->SUPER::prop(@_) if ref $_[0] eq 'HASH';
	
	my $val = $self->SUPER::prop(@_);
	return $val if defined $val;

	my $name = shift;
	return $_[0] if @_;

	my $window = $parathia{$self};

	my $ret =
		$name =~ /^(?:0|[1-9]\d*)\z/ && $name < 4294967295
        	? $window->frames->[$name]
		: $window->frames->{$name};
	defined $ret ? $self->upgrade($ret) : $ret;
}

sub set {
	my $obj = shift;
	my $val = pop;
	croak "Not enough arguments for W:M:P:JS:JE->set" unless @_;
	my $prop = pop;
	for (@_) {
		my $next_obj = $obj->{$_};
		defined $next_obj or
			$obj->{$_} = {},
			$obj = $obj->{$_}, next;
		$obj = $next_obj;
	}
	$obj->{$prop} = $val;
	return;
}

sub bind_classes {
	my($self, $classes) = @_;
	my @defer;
	for (grep /::/, keys %$classes) {
		my $i = $$classes{$$classes{$_}}; # interface info
		my @args = (
			unwrap  => 1,
			package => $_,
			name    => $$classes{$_},
			methods => { map {
			   my $perlname = $_;
			   $perlname .= 16 if $$i{$_} & UTF16;
			   $_ => $$i{$_} & VOID
			    ? $perlname
			    : "$perlname:$types[$$i{$_} & TYPE]"
			  } grep !/^_/ && $$i{$_} & METHOD, keys %$i },
			props => { map {
			   my $perlname = $_;
			   $perlname .= 16 if $$i{$_} & UTF16;
			   $$i{$_} & READONLY
			      ? ($_ =>{
			          fetch=>
			           "$perlname:$types[$$i{$_} & TYPE]"
			        })
			      : ($_ => "$perlname:$types[$$i{$_} & TYPE]")
			 } grep !/^_/ && !($$i{$_} & METHOD), keys %$i  },
			hash  => $$i{_hash},
			array => $$i{_array},
			exists $$i{_isa} ? (isa => $$i{_isa}) : (),
			exists $$i{_constructor}
				? (constructor => $$i{_constructor})
				: (),
		);
		my $make_constants;
		if(exists $$i{_constants}){
		  my $p = $_;
		  $make_constants = sub { for(@{$$i{_constants}}){
			/([^:]+\z)/;
			$self->{$$classes{$p}}{$1} =
			# ~~~ to be replaced simply with 'eval' when JE's
			#     upgrading is improved:
				$self->upgrade(eval)->to_number;
		}}}
		if (exists $$i{_isa} and !exists $self->{$$i{_isa}}) {
			push @defer, [\@args, $$i{_isa}, $make_constants]
		} else {
#			use Data::Dumper; print Dumper \@args if $_ !~ /HTML|CSS/;
			$self->bind_class(@args);
			defined $make_constants and &$make_constants;
		}
	}
	while(@defer) {
		my @copy = @defer;
		@defer = ();
		for (@copy) {
			if(exists $self->{$$_[1]}) { # $$_[1] == superclass
				$self->bind_class(@{$$_[0]});
				&{$$_[2] or next}
			}
			else {
				push @defer, $_;
			}
		}
	}
	return;
}

sub event2sub {
	my ($w, $code, $elem, $url, $line) = @_;

	# ~~~ JE's interface needs to be improved. This is a mess:
	# We need the line break after $code, because there may be a sin-
	# gle-line comment at the end,  and no line break.  ("foo //bar"
	# would fail without  this,  because  the  })  would  be  com-
	# mented out too.)
	# We have to check whether the $elem is a form before calling it’s
	#‘form’  method,  because forms *do*  have such a method,  but it
	# returns a list of form element names and values, which is *not*
	# what we  want.  (We  want  the  element’s  parent  form  where
	# applicable.)
	($w->compile("(function(){ $code\n })",$url,$line)||return)
		->execute($w, bless [
			$w,
			$w->upgrade($elem->ownerDocument),
			$elem->tag ne 'form' && $elem->can('form')
			  ? $w->upgrade($elem->form) : (),
			my $wrapper=($w->upgrade($elem))
		], 'JE::Scope');
}

sub define_setter {
	my $obj = shift;
	my $cref = pop;
	my $prop = pop;
	for (@_) {
		my $next_obj = $obj->{$_};
		defined $next_obj or
			$obj->{$_} = {},
			$obj = $obj->{$_}, next;
		$obj = $next_obj;
	}
	$obj->prop({name=>$prop, store=>sub{$cref->($_[1])}});
	return;
}

sub new_function {
	my($self, $name, $sub, $type) = @_;
	if(defined $type) {
		$self->new_function(
			$name => sub { $self->{$type}->($sub->(@_)) }
		);
		weaken $self;
	} else {
		shift->SUPER::new_function(@_);
	}
}

sub exists {
	my $self = shift;
	SUPER::exists $self @_
	 or do {
		my($name) = @_;
		my $window = $parathia{$self};
		$name =~ /^(?:0|[1-9]\d*)\z/ && $name < 4294967295
		? defined $window->frames->[$name]
		: defined $window->frames->{$name}
	}
}

sub destroy {
 return SUPER::destroy{@_} if caller eq 'JE::Destroyer';
 JE'Destroyer'destroy $_[0]
}


package WWW::Scripter::Plugin::JavaScript::JE::Proxy;

# ~~~ This package needs to override any method used internally by
#     JavaScript that returns a JE datatype to return a wrapper instead.

sub new {
 # If a method is called on another window, as in frames[0].alert(), an
 # lvalue is created with the proxy object returned by frames[0]  as the
 # the base and ‘alert’ as the property name. The lvalue then fetches the
 # value of frames[0].alert  and runs it with  frames[0]  as the invocant.
 # This means that it is the proxy  that  gets  passed  to  the  closures
 # in  WSPJSJE:new  above.  Since  those  expect  to  find  a  window  in
 # $parathia{$self}, and $self might be a proxy, we have to add ourselves
 # to the %parathi field hash and store the window there. For the sake of
 # speed in autoload, we still keep a direct  reference  to  the  window.
 my $proxy = bless \(my $w = pop);
 Scalar'Util'weaken( $parathia{$proxy} = $w );
 $proxy
}

sub AUTOLOAD {
 my $window = ${;shift};
 (
   # We have to use this roundabout method
   # rather than __PACKAGE__->new($window),
   # because the JS plugin needs to do its
   # stuff (binding classes, etc.).
   $window->plugin("JavaScript")->back_end($window)
 )->${\(our $AUTOLOAD =~ /.*::(.+)\z/)[0]}(@_)
}

sub DESTROY{}
sub isa {
 goto &UNIVERSAL'isa;
}


=cut

# ------------------ DOCS --------------------#

=head1 NAME

WWW::Scripter::Plugin::JavaScript::JE - JE backend for WSPJS

=head1 VERSION

0.008 (alpha)

=head1 DESCRIPTION

This little module is a bit of duct tape to connect the JavaScript plugin
for L<WWW::Scripter> to the JE JavaScript engine. Don't use this module
directly. For usage, see
L<WWW::Scripter::Plugin::JavaScript>.

=head1 REQUIREMENTS

Hash::Util::FieldHash::Compat

HTML::DOM 0.032 or later

JE 0.056 or higher

WWW::Scripter 0.016 or higher

=head1 SEE ALSO

=over 4

=item -

L<WWW::Scripter::Plugin::JavaScript>

=item -

L<JE>

=item -

L<HTML::DOM>

=item -

L<WWW::Mechanize::Plugin::JavaScript::JE> (the original version of this
module)

=cut
